### PR TITLE
logind: switch sessions_by_leader to PidRef

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/login/logind-session.c [hunk 1]
```diff
@@ -180,14 +191,17 @@ int session_set_leader_consume(Session *s, PidRef _leader) {
         if (pidref_equal(&s->leader, &pidref))
                 return 0;
 
-        r = hashmap_put(s->manager->sessions_by_leader, PID_TO_PTR(pidref.pid), s);
+        r = pidref_dup(&pidref, &pid_dup);
```
Comment (yuwata) [inline src/login/logind-session.c]: Why not to directly use `s->leader`?
I know `pidref_hash_ops` frees the PidRef, but I think in this case it is better to rename `pidref_hash_ops` -> `pidref_hash_ops_free`, and re-introduce `pidref_hash_ops` without destructor, and use the new hash ops here.

File: src/login/logind-session.c [hunk 2]
```diff
@@ -87,6 +87,19 @@ int session_new(Session **ret, Manager *m, const char *id) {
...
+        _cleanup_(pidref_freep) PidRef *pid_old = NULL;
+
+        assert(s);
+
+        if (!pidref_is_set(&s->leader))
+                return;
+
+        assert_se(hashmap_remove2(s->manager->sessions_by_leader, &s->leader, (void**) &pid_old) == s);
+
+        return pidref_done(&s->leader);
```
Comment (yuwata) [inline src/login/logind-session.c:98]: In this stage, `s->leader` and `pid_old` points to the same object, right? Hmm...

**Review by yuwata — CHANGES_REQUESTED** [review]

**Review by yuwata — APPROVED** [review]

**Comment by YHNdnzj** [comment]
Tweaked `manager_get_session_by_pidref` further so that it actually verifies the passed `PidRef`. PTAL.

File: src/login/logind-core.c [hunk 3]
```diff
@@ -734,8 +742,8 @@ int manager_read_utmp(Manager *m) {
                 if (isempty(t))
                         continue;
 
-                s = hashmap_get(m->sessions_by_leader, PID_TO_PTR(u->ut_pid));
-                if (!s)
+                r = manager_get_session_by_pidref(m, &PIDREF_MAKE_FROM_PID(u->ut_pid), &s);
+                if (r <= 0)
```
Comment (yuwata) [inline src/login/logind-core.c]: `r` is not used later, hence the lines can be shorten by `if (manager_get_session_by_pidref(...) <= 0)`.
Feel free to ignore if you do not like the form.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/core/unit.c [hunk 4]
```diff
@@ -2944,7 +2944,7 @@ void unit_unwatch_pidref(Unit *u, PidRef *pid) {
 
         if (uu == u)
                 /* OK, we are in the first table. Let's remove it there then, and we are done already. */
-                assert_se(hashmap_remove_value(u->manager->watch_pids, pid2, uu) == uu);
+                assert_se(hashmap_remove_value(u->manager->watch_pids, pid2, uu));
```
Comment (poettering) [inline src/core/unit.c:2947]: hmm, why drop this? this is a safety check that the entry matches what we expect here.

Comment (poettering) [reply src/core/unit.c:2947]: ah, i see why that's unnecessary.

Comment (YHNdnzj) [reply src/core/unit.c:2947]: Please check the commit message - the only difference between `hashmap_remove` and `_remove_value` is that the latter does the check internally, and returns NULL if things don't match IIUC?

Comment (yuwata) [reply src/core/unit.c:2947]: It always return `uu` or `NULL`, and `NULL` is avoided by the assertion. So, I think this change is reasonable.

**Comment by poettering** [comment]
lgtm

**Pull request merged**